### ISSUE TITLE
Redirection to a different host

### ISSUE THREAD

**Issue by craigfrancis** [mozilla/http-observatory#226]
I believe this is related to the TODO on:

https://github.com/mozilla/http-observatory/blob/master/httpobs/scanner/analyzer/misc.py#L167

---

If you have a website that does a redirect, such as:

    http://www.example.com
    https://example.com
    https://www.example.com

In some cases it seems to be fine:

https://observatory.mozilla.org/analyze.html?host=www.livingworduk.org

    "All hosts redirected to are in the HTTP Strict Transport Security (HSTS) preload list"

    Scan: 3570373

    curl -L -v http://www.livingworduk.org 2>&1 | grep Location
    < Location: https://livingworduk.org/
    < Location: https://www.livingworduk.org/

In other cases, it complains:

https://observatory.mozilla.org/analyze.html?host=www.odysseyconservationtrust.com

    "Initial redirection from http to https is to a different host, preventing HSTS"

    Scan: 3570371

    curl -L -v http://www.odysseyconservationtrust.com 2>&1 | grep Location
    < Location: https://odysseyconservationtrust.com/
    < Location: https://www.odysseyconservationtrust.com/

As far as I can tell, these websites (which are quite old) have the same Apache config (with the exception of the domain name itself), they are running on the same server, and both domains have been HSTS preloaded.

**Comment by april** [comment]
The Observatory uses a local copy of the preload list for performance and CI reasons.  A new version was recently committed that added odysseyconversationtrust.com, and should hopefully go up sometime in the next week or two.  :)

**Comment by craigfrancis** [comment]
That would explain why the two websites have different results... but it the error message is incorrect at the moment.

:-)

i.e. the second website, which isn't in your HSTS preload list yet, is doing the redirects correctly, but gets the error message:

    "Initial redirection from http to https is to a different host, preventing HSTS"

It's on the same host, it's just dropping the "www" sub-domain first (required for HSTS preload, and so browsers will correctly set the HSTS for the whole domain).

Where I believe it's just down to:

https://github.com/mozilla/http-observatory/blob/master/httpobs/scanner/analyzer/misc.py#L167

:-)

**Comment by april** [comment]
Yeah, it's simply detecting if the first redirection is off host and not checking to see if it gets sent back there eventually.  If you went from: http://www.odysseyconservationtrust.com -> https://www.odysseyconservationtrust.com -> https://odysseyconservationtrust.com, it would be fine.

It probably could be updated to detect two things:

* Being sent off domain and then back again
* Being sent to a higher level domain that sets HSTS with includeSubDomains

**Comment by craigfrancis** [comment]
Yes please, especially the second point :-)

As to that redirect chain... it would keep the Observatory validator happy, but it would upset Lucas (hstspreload.org), as he want's the browser to go to the non-www domain first, so the browser will apply HSTS to the whole domain (ref `includeSubDomains`):

https://github.com/chromium/hstspreload.org/issues/33

And I'd be a little unhappy about using the non-www version of the domain... because that's only "right and proper" (not really, it just helps with keeping cookies locked to that www subdomain).

http://www.yes-www.org/why-use-www/

Thanks again April... it's not a major issue for me (as I try to preload all my domains), but it would be nice to get resolved :-)

**Comment by floatingatoll** [comment]
I filed mozilla/http-observatory-website#107 "HSTS-safe redirect chain visualizer", because I don't know of any tools anywhere that let people input a domain name and get back a series of HSTS-compliant redirects to implement.

**Comment by april** [comment]
Just as an aside, I upgraded the scanners today and it looks like everything is working with your other domain with regards to preloading.  :)

**Comment by craigfrancis** [comment]
Thanks April :-)

**Comment by luigigubello** [comment]
Hi :)
Is this issue fixed?
